### ISSUE TITLE
T5Gemma support

### ISSUE THREAD

**Issue by betweenus** [google/gemma.cpp#661]
No program supports T5Gemma model. Please could you add its support here or better in llama.cpp?

**Comment by jan-wassenberg** [comment]
Hi, we did not yet have plans to add it here but have added it to the wishlist. No telling when someone will get to it. Would certainly speed things up if someone were to join in and send patches, which are welcome :)

**Comment by construimosfuturoluli-create** [comment]
claro

**Comment by maasir554** [comment]
Hi @jan-wassenberg I would like to work on this one, will it still be accepted?

**Comment by jan-wassenberg** [comment]
Hi @maasir554 , yes, it would be welcome :) Would advise searching for all occurrences of a model enum such as GEMMA3_4B, and adding something similar. Please also target the dev branch.

**Comment by maasir554** [comment]
Hi @jan-wassenberg , thanks for the response! 
Apologies for the delay getting back to this. I am actively back on it now.

I am currently working on converting the T5Gemma `.safetensors` to the `.sbs` format. I saw that a pipeline already exists for `PaliGemma` in /python/ (relying on /compression/), so my plan is to fork that logic for T5.

Note: I ran into a few minor Bazel and config issues on the latest dev branch (like the removal of the `Model::PALLIGEMMA_448` enum in /python and `//third-party/bazel_rules` in /compression/python/BUILD.bazel), but I am working through them.

To keep the repository clean, I plan to place the new conversion script in the existing /python/ directory (e.g., /python/convert_t5.py). Please let me know if you would prefer a different structure for this!

**Comment by jan-wassenberg** [comment]
Hi, no worries :)
Yes, python/convert_from_safetensors.py is currently what we have. No progress yet on open sourcing our pipeline for Gemma. Would be nice to integrate into this same file, rather than to fork a copy, because copies bit-rot and they anyway share lots of code. We can imagine just skipping the vision parts.
Happy to advise on the current issues if you hit a roadblock :)

**Comment by maasir554** [comment]
Hi @jan-wassenberg , thanks for the direction!
quick note: while smoke-testing my branch on macOS Apple Silicon, I hit a build failure on latest `dev` in `gemma/flash_attention.cc`.

this reproduces even before my T5Gemma runtime changes, so I don't think it's caused by my branch. I checked by building the commit before my runtime foundation work, and it failed in the same area.

Environment:
- macOS on Apple Silicon
- branch synced with latest `upstream/dev`
- failing target: `bazel build //:gemma`

The error is in Highway ARM NEON code, around `InterleaveLower` / `InterleaveUpper` instantiation from `flash_attention_arm-inl.h`.

Relevant failure shape:

```txt
gemma/flash_attention.cc failed
external/highway~/hwy/ops/arm_neon-inl.h:6526:10: error: no matching function for call to 'InterleaveLower'
./gemma/flash_attention_arm-inl.h:660:25: note: instantiation ...
N_NEON_BF16 / N_NEON_WITHOUT_AES
```
I tried disabling all NEON targets with:
```
--copt=-DGEMMA_DISABLED_TARGETS='(HWY_SCALAR|HWY_ALL_NEON|HWY_SVE|HWY_SVE2|HWY_SVE_256)'
```
That allowed `//:gemma` to build using the fallback path, but it was too slow for useful model smoke testing.

Do you know if this is a known issue with the current Highway / FlashAttention path on Apple Silicon, or is there a preferred Bazel flag / config for local smoke tests on this platform?

**Comment by maasir554** [comment]
Quick update: I now have an initial end-to-end T5Gemma S/S path working locally on top of `python/convert_from_safetensors.py` and the gemma.cpp runtime.

Current status:
- Converted HF `.safetensors` T5Gemma S/S weights to `.sbs`
- Loads the generated `.sbs` in gemma.cpp
- Runs fresh seq2seq generation from encoder input + BOS decoder start
- BF16 conversion is the default path for correctness
- Basic optimized local smoke test runs successfully

local smoke test I ran:
- model: `t5gemma-s-s-it`
- prompt: `Hello`
- output: `I am a new user, ...`
- opt build speed: roughly **30-35 generated tok/s** on my machine (Macbook M1, 8gb RAM)

Limitations:
- Initial PR targets T5Gemma v1 S/S only
- T5Gemma 2 is currently out of scope
- Runtime supports fresh seq2seq generation only for now, not multiturn/cache reuse
- T5 attention/logits paths are functional but not fully optimized like existing decoder-only Gemma paths

Link to my working branch: https://github.com/maasir554/gemma.cpp/tree/t5gemma-support 

Shall I now open a PR? Please let me know your thoughts.

**Comment by jan-wassenberg** [comment]
Congrats, glad to hear you have it running :)
If I understood correctly, you have a compile error and it only works without NEON/SVE?
We aren't seeing the flash attention build failure, I wonder if you have any custom compiler flags, in particular anything `-m`?

**Comment by maasir554** [comment]
Thanks! You got it right, but this error is not blocking my T5Gemma PR anymore, so I’ll keep focusing there for now. Happy to revisit this separately if it becomes useful. 
For context, I don’t think I have custom `-m...` flags set, but I will investigate separately if needed.

**Comment by jan-wassenberg** [comment]
Got it, thanks. I would be curious to understand the issue because it might affect others as well.
I wonder if it's an older compiler/Xcode command line tool version?

**Comment by maasir554** [comment]
Thanks, that makes sense. I checked the local toolchain versions:

macOS: 26.5.2 (25F84)
Xcode: 26.4 (17E192)
xcode-select: /Applications/Xcode.app/Contents/Developer
Apple clang: 21.0.0 (clang-2100.0.123.102)
Target: arm64-apple-darwin25.5.0
Bazel: 7.1.1

So it does not look like an older Xcode/CLT version.

The failing command for gemma/flash_attention.cc does not seem to include custom -march/-mcpu style flags. The only -m... flag I see is: `-mmacosx-version-min=26.4`

The compile action is roughly:
```text
external/rules_cc~~cc_configure_extension~local_config_cc/cc_wrapper.sh
  -std=c++17
  -D_USE_TF_STRING_VIEW
  -mmacosx-version-min=26.4
  ...
  -c gemma/flash_attention.cc
  -o bazel-out/darwin_arm64-fastbuild/bin/_objs/attention/flash_attention.o
```

The failure still reproduces and is in the NEON BF16 path:
```
external/highway~/hwy/ops/arm_neon-inl.h:6526:10:
error: no matching function for call to 'InterleaveLower'

./gemma/flash_attention_arm-inl.h:660:25:
note: in instantiation of function template specialization
'hwy::N_NEON_BF16::InterleaveLower<...>' requested here

gemma/flash_attention.cc:1588:7:
note: in instantiation of function template specialization
'gcpp::N_NEON_BF16::TileFlashAttentionReturnExpSumsAndMaxLogitsBF16<...>'
requested here
```

There is also a similar `InterleaveUpper` failure later from `arm_neon-inl.h:6533`.
So the issue still reproduces locally with current Xcode/clang, and I don’t see custom arch flags in the failing compile command.
